components/collator/src/comparison.rs

```diff
@@ -100,26 +105,37 @@ impl Collator {
             // the default, the code here that omits the explicit
             // variant if it matches the default should become
             // unnecessary.
-            let original_locale = locale;
-            let mut filtered_locale = alloc::borrow::Cow::Borrowed(locale);
-
-            let key = icu_locid::extensions_unicode_key!("co");
-            if let Some(variant) = original_locale.get_unicode_ext(&key) {
-                let zh = filtered_locale.language() == icu_locid::subtags_language!("zh");
-                let sv = filtered_locale.language() == icu_locid::subtags_language!("sv");
+            const KEY: Key = key!("co");
+            if let Some(variant) = locale.get_unicode_ext(&KEY) {
+                let zh = locale.language() == language!("zh");
+                let sv = locale.language() == language!("sv");
                 // Omit the explicit collation variant if it is the default.
                 // "standard" is the default for all languages except zh and sv.
-                if !((!zh && !sv && variant == value!("standard"))
+                if (!zh && !sv && variant == value!("standard"))
                     || (zh && variant == value!("pinyin"))
-                    || (sv && variant == value!("reformed")))
+                    || (sv && variant == value!("reformed"))
                 {
-                    filtered_locale.to_mut().set_unicode_ext(key, variant);
+                    // Remove -u-co- and all other extensions by cloning just the langid
+                    holder = DataLocale::from(locale.get_langid());
+                    &holder
+                } else {
+                    // Remove all other extensions by cloning the langid and adding -u-co-
+                    let mut locale = Locale::from(locale.get_langid());
+                    locale.extensions.unicode.keywords = Keywords::new_single(KEY, variant);
+                    holder = DataLocale::from(locale);
+                    &holder
                 }
+            } else if locale.has_unicode_ext() {
+                // Remove all extensions by cloning the langid
```

Comment:
Thought: If you have vertical fallback enabled, this should be handled for you already; I think it's only a problem if you are accessing data without a vertical fallback adapter. I've long wanted to have a centralized solution to the "unnecessary unicode extension keywords" problem for non-fallback users but we don't have one at this time because the fallback adapter is the only thing we promise to work. With compile-time fallbacking coming in 1.3, we'll need to think about this more.

Reply:
The irrelevant extension subtags are removed here:

https://github.com/unicode-org/icu4x/blob/3d73bbccf27e5f99bbd40b71c8f02a93099ec42c/provider/adapters/src/fallback/algorithms.rs#L55

Note that I would consider it not exactly correct to pluck them in the Collator constructor (and I think why I changed the code to work this way) because there could be child constructors that need different extension keys. For example, if Collator needed to pull in FixedDecimalFormatter for some reason, it would be wrong to remove the -u-nu subtag. Or, for a more concrete example, if Normalizer got its own extension key, we shouldn't remove it.

Reply:
In other words, the current policy, which we should document better, is, you have two options:

1. Use the vertical fallback adapter, or
2. Know what you are doing and pass in raw locales that go directly to data lookup.

Technically there is a third option, which is the vertical fallback adapter constructed without data: it will at least normalize the locale extension keywords for you.

Reply:
At what point do default extensions get removed?

Reply:
The default extensions get removed while fallback is running, like

en-US-u-co-standard
en-US
en
und